### PR TITLE
Add open-drain GPIO support.

### PR DESCRIPTION
**Pull request by tzarc**
## Description

Adds support for open-drain outputs.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by drashna — APPROVED** [review]

File: platforms/avr/gpio.h [hunk 1]
```diff
@@ -25,7 +25,9 @@ typedef uint8_t pin_t;
 #define setPinInput(pin) (DDRx_ADDRESS(pin) &= ~_BV((pin)&0xF), PORTx_ADDRESS(pin) &= ~_BV((pin)&0xF))
 #define setPinInputHigh(pin) (DDRx_ADDRESS(pin) &= ~_BV((pin)&0xF), PORTx_ADDRESS(pin) |= _BV((pin)&0xF))
 #define setPinInputLow(pin) _Static_assert(0, "AVR processors cannot implement an input as pull low")
-#define setPinOutput(pin) (DDRx_ADDRESS(pin) |= _BV((pin)&0xF))
+#define setPinOutputPushPull(pin) (DDRx_ADDRESS(pin) |= _BV((pin)&0xF))
+#define setPinOutputOpenDrain(pin) _Static_assert(0, "AVR platform does not implement an open-drain output")
```
Comment (sigprof) [inline platforms/avr/gpio.h:29]: Technically it is possible to implement an open-drain output on AVR — set the PORTx bit to 0, then write the inverted output value to the DDRx bit. The problem is that it requires either some conditional code in the `writePin()` implementation, or a separate set of macros like `writeOpenDrainPin()`, `writeOpenDrainPinHigh()`, `writeOpenDrainPinLow()` (with bad consequences if you use a wrong kind of macro for the chosen pin mode).

Comment (tzarc) [reply platforms/avr/gpio.h:29]: Yep, but nobody has asked for it, and nobody's using it. Figured it'd be a problem for another day.

File: platforms/chibios/gpio.h [hunk 2]
```diff
@@ -41,10 +43,10 @@ typedef uint16_t port_data_t;
...
 #define readPort(pin) palReadPort(PAL_PORT(pin))
 
-#define setPortBitInput(pin, bit) palSetPadMode(PAL_PORT(pin), bit, PAL_MODE_INPUT)
-#define setPortBitInputHigh(pin, bit) palSetPadMode(PAL_PORT(pin), bit, PAL_MODE_INPUT_PULLUP)
-#define setPortBitInputLow(pin, bit) palSetPadMode(PAL_PORT(pin), bit, PAL_MODE_INPUT_PULLDOWN)
-#define setPortBitOutput(pin, bit) palSetPadMode(PAL_PORT(pin), bit, PAL_MODE_OUTPUT_PUSHPULL)
+#define setPortBitInput(pin, bit) palSetPadMode(PAL_PORT(pin), (bit), PAL_MODE_INPUT)
+#define setPortBitInputHigh(pin, bit) palSetPadMode(PAL_PORT(pin), (bit), PAL_MODE_INPUT_PULLUP)
+#define setPortBitInputLow(pin, bit) palSetPadMode(PAL_PORT(pin), (bit), PAL_MODE_INPUT_PULLDOWN)
+#define setPortBitOutput(pin, bit) palSetPadMode(PAL_PORT(pin), (bit), PAL_MODE_OUTPUT_PUSHPULL)
```
Comment (sigprof) [inline platforms/chibios/gpio.h:49]: Hmm, looks like the implementation of `setPortBitOutputPushPull()` and `setPortBitOutputOpenDrain()` is missing here?

Comment (tzarc) [reply platforms/chibios/gpio.h]: As discussed on Discord, given that it's not in use anywhere but one keyboard, and it has its own implementation of the port-based GPIO functions, the port-related operations have been removed for the time being until we work out a proper solution for them.

**Pull request merged**